flang/lib/Lower/OpenMP/DataSharingProcessor.cpp

```diff
@@ -344,7 +353,6 @@ void DataSharingProcessor::collectSymbols(
     assert(curScope && "couldn't find current scope");
     if (isPrivatizable(*sym) && !symbolsInNestedRegions.contains(sym) &&
         !privatizedSymbols.contains(sym) &&
-        !sym->test(Fortran::semantics::Symbol::Flag::OmpPreDetermined) &&
```

Comment:
This test was preventing induction variables, when `default(private)` is specified, from being added to the list of symbols privatized by the `DataSharingProcessor`.

Reply:
I've added this test to make `defaultSymbols` and `implicitSymbols` more accurate and to simplify `defaultPrivatize` and `implicitPrivatize`.

Also, implicit symbols (those whose DSA is implicitly determined) are only present when there are no default symbols in a given construct (see`collectImplicitSymbols`). If pre-determined symbols are included in `defaultSymbols` we can no longer know what is the set of default symbols.

Finally, in the non-delayed privatization path, pre-determined symbols were already being skipped:
```c++
void DataSharingProcessor::cloneSymbol(const Fortran::semantics::Symbol *sym) {
  // Privatization for symbols which are pre-determined (like loop index
  // variables) happen separately, for everything else privatize here.
  if (sym->test(Fortran::semantics::Symbol::Flag::OmpPreDetermined))
    return;
```

As pre-determined symbols' DSAs don't depend on default clauses, I don't think they should be handled as default or implicit symbols. In your test program, if you remove `default(private)`, are induction variables correctly privatized?

Reply:
> Finally, in the non-delayed privatization path, pre-determined symbols were already being skipped:

Indeed. And this is what I am trying to prevent from happening here both for early and delayed privatization. AFAIU, I don't think there is a reason to have 2 different code-gen paths for pre-determined vs. non-pre-determined variables. Please let me know if I am mistaken here.

More context for this is in this Slack thread: https://flang-compiler.slack.com/archives/C01PY03PP9P/p1714729085154889.

>  if you remove default(private), are induction variables correctly privatized?

In that case, they are not privatized at all in the sense that their alloca's are emitted simply as part of emitting the code for the do concurrent for which they are used as induction variables. They are not marked as pre-determined.

Reply:
> more accurate and to simplify defaultPrivatize and implicitPrivatize.

Can you elaborate on what you mean by simplify?

Reply:
> > Finally, in the non-delayed privatization path, pre-determined symbols were already being skipped:
> 
> Indeed. And this is what I am trying to prevent from happening here both for early and delayed privatization. AFAIU, I don't think there is a reason to have 2 different code-gen paths for pre-determined vs. non-pre-determined variables. Please let me know if I am mistaken here.
> 

I also don't think there is a reason to have 2 different paths to perform the privatization of pre-determined vs. non-predetermined variables. However, currently `DataSharingProcessor::collectSymbols` is intended to collect default and implicit symbols only. It won't even be called if there are no default or implicit symbols.

> More context for this is in this Slack thread: https://flang-compiler.slack.com/archives/C01PY03PP9P/p1714729085154889.
> 
> > if you remove default(private), are induction variables correctly privatized?
> 
> In that case, they are not privatized at all in the sense that their alloca's are emitted simply as part of emitting the code for the do concurrent for which they are used as induction variables. They are not marked as pre-determined (see [this](https://flang-compiler.slack.com/archives/C01PY03PP9P/p1714899066421309?thread_ts=1714729085.154889&cid=C01PY03PP9P)).

Thanks for pointing this out. I think this is the main problem. Pre-determined (private) variables must be privatized independently of the presence of `private` or `default` clauses. So there is something wrong with case 2, as the loop variable should also be marked as pre-determined in semantics phase.

But actually, it seems that currently loop variables are always being assumed to be pre-determined, which, AFAIU, is not always the case (see test/Semantics/OpenMP/symbol08.f90, for instance).

Right now it would probably be easier to refactor `genLoopVars`, to somehow pass the loop variables to the DataSharingProcessor, so that privatization happens only there.

Reply:
> > more accurate and to simplify defaultPrivatize and implicitPrivatize.
> 
> Can you elaborate on what you mean by simplify?

Before #85989, `defaultSymbols` would have other unrelated symbols, that would be filtered at `defaultPrivatize` only. The filter part was moved to `collectSymbols`, which simplified `defaultPrivatize` and `implicitPrivatized` and allowed `symbolsInNestedRegions` to be removed.

Reply:
Thanks for the clarification. I understand the context for that change better now.

> Right now it would probably be easier to refactor genLoopVars, to somehow pass the loop variables to the DataSharingProcessor, so that privatization happens only there.

I have a different proposal which I implemented in the latest commit on the PR. To keep the different lists of symbols (default, implicit, explicit) accurate, I created one more list of symbols to aggregate all symbols to be privatized including the pre-determined ones. And then, we would have a single `privatize` method to go over the aggregated list.

The problem with invoking the `DataSharingProcessor` from `genLoopVars` is that I think we will have to instantiate it from `genLoopVars`. I think instantiating and controlling instances of the DSP should be localized to the OMP constructs that need it in `Lower/OpenMP.cpp`.

Let me know if you disagree with the current approach.